### PR TITLE
Group AccountReset extensions, implement full Account Reset flow

### PR DESCRIPTION
**Pull request by Cheddam**
This PR introduces the post-email section of the Account Reset flow. It also groups the Account Reset related classes into a separate folder within the Extension namespace - keen to get feedback on this, as it muddies the namespace structure a little.

**TODO:**

- [x] Improve test coverage
- [x] Useful error responses for users

### CODE REVIEW

File: src/Extension/AccountReset/MemberExtension.php [hunk 1]
```diff
@@ -65,16 +65,17 @@ public function generateAccountResetTokenAndStoreHash(): string
      * @param string $token
      * @return Member
```
Comment (robbieaverill) [inline src/Extension/AccountReset/MemberExtension.php]:
```suggestion
     * @return bool
```

File: src/Extension/AccountReset/SecurityAdminExtension.php [hunk 2]
```diff
@@ -147,10 +149,9 @@ protected function sendResetEmail($member)
      * @param Member $member
      * @param string $token
      * @return string
-     * @todo Implement when Account Reset Handler is built
      */
     protected function getAccountResetLink(Member $member, string $token): string
     {
-        return '';
+        return Security::singleton()->Link('resetaccount') . "?m={$member->ID}&t={$token}";
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityAdminExtension.php]:
```suggestion
        return Controller::join_links(
            Security::singleton()->Link('resetaccount'),
            "?m={$member->ID}&t={$token}"
        );
```

File: src/Extension/AccountReset/SecurityExtension.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ * Extends the Security controller to support Account Resets. This extension can
+ * itself be extended to add procedures to the reset action (such as removing
+ * additional authentication factors, sending alerts, etc.)
+ *
+ * @package SilverStripe\MFA\Extension
+ * @property Security owner
+ */
+class SecurityExtension extends Extension
+{
+    use JSONResponse, BaseHandlerTrait, Extensible;
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: I think linting will tell you to separate these into individual `use` statement

File: src/Extension/AccountReset/SecurityExtension.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    private static $url_handlers = [
+        'GET reset-account' => 'resetaccount',
+        'ResetAccountForm' => 'ResetAccountForm',
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: Is this line necessary?

File: src/Extension/AccountReset/SecurityExtension.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        }
+
+        ['m' => $memberID, 't' => $token] = $request->getVars();
+
+        /** @var Member|MemberExtension $member */
+        $member = Member::get()->byID($memberID);
+
+        if (is_null($member) || $member->verifyAccountResetToken($token) === false) {
+            return $this->jsonResponse(['error' => 'Invalid member or token'], 400);
+        }
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: Do you need to enforce sudo mode for this action?

File: src/Extension/AccountReset/SecurityExtension.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+     * @param array $data
+     * @param Form $form
+     * @return HTTPResponse
+     */
+    public function doResetAccount(array $data, Form $form): HTTPResponse
+    {
+        $memberID = $this->owner->getRequest()->getSession()->get('MemberID');
+
+        // If the ID isn't in the session, politely assume the session has expired
+        if (!$memberID) {
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: Do we need sudo mode and CSRF protection here?

File: src/Extension/AccountReset/SecurityExtension.php (hunk 5 above)
Comment (Cheddam) [reply src/Extension/AccountReset/SecurityExtension.php]: I don't think sudo mode _can_ be enforced during an account reset - you don't have a password to provide whilst you're resetting your password.

File: src/Extension/AccountReset/SecurityExtension.php (hunk 6 above)
Comment (Cheddam) [reply src/Extension/AccountReset/SecurityExtension.php]: See above comment about sudo mode. CSRF protection is enabled by default on Forms (I've tested this as working on Account Reset).

File: src/Extension/AccountReset/SecurityExtension.php [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        'ResetAccountForm',
+    ];
+
+    public function resetaccount(HTTPRequest $request): HTTPResponse
+    {
+        if (Security::getCurrentUser()) {
+            return $this->jsonResponse(['error' => 'Already authenticated'], 400);
+        }
+
+        ['m' => $memberID, 't' => $token] = $request->getVars();
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: This syntax confuses me, what's it doing?

Comment (ScopeyNZ) [reply src/Extension/AccountReset/SecurityExtension.php]: Yeah this is new in PHP 7.1.

File: src/Extension/AccountReset/SecurityExtension.php [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                    "The account reset process timed out. Please click the link in the email and try again."
+                ),
+                ValidationResult::TYPE_ERROR
+            );
+
+            return $this->owner->redirectBack();
+        }
+
+        /** @var Member&MemberExtension $member */
+        $member = Member::get()->byID(intval($memberID));
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: I think generally `(int) $memberID` would be preferable (non blocking comment)

File: src/Extension/AccountReset/SecurityExtension.php [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+        // Fail if passwords do not match
+        if ($data['NewPassword1'] !== $data['NewPassword2']) {
+            $form->sessionMessage(
+                _t(
+                    'SilverStripe\\Security\\Member.ERRORNEWPASSWORD',
+                    'You have entered your new password differently, try again'
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: This renders a little weird in login forms. Might be my fault from my last PR there though...

image

Also, are you sure about this wording? It doesn't sit perfectly with me.

File: src/Extension/AccountReset/SecurityExtension.php [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        $request->getSession()->set('MemberID', $member->ID);
+
+        return $this->owner->getResponse()->setBody($this->owner->renderWith(
+            'Security',
+            [
+                'Title' => _t(
+                    __CLASS__ . '.ACCOUNT_RESET_TITLE',
+                    'Reset Account'
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]: I'd suggest a minor case change:

```suggestion
                    'Reset account'
```

File: src/Extension/AccountReset/SecurityExtension.php [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    public function ResetAccountForm(): Form
+    {
+        $fields = FieldList::create([
+            PasswordField::create(
+                'NewPassword1',
+                _t(
+                    'SilverStripe\\Security\\Member.NEWPASSWORD',
+                    'New Password'
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]:
```suggestion
                    'New password'
```

File: src/Extension/AccountReset/SecurityExtension.php [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+                )
+            ),
+            PasswordField::create(
+                'NewPassword2',
+                _t(
+                    'SilverStripe\\Security\\Member.CONFIRMNEWPASSWORD',
+                    'Confirm New Password'
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]:
```suggestion
                    'Confirm new password'
```

File: src/Extension/AccountReset/SecurityExtension.php [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                )
+            ),
+        ]);
+
+        $actions = FieldList::create([
+            FormAction::create('doResetAccount', 'Reset Account'),
```
Comment (robbieaverill) [inline src/Extension/AccountReset/SecurityExtension.php]:
```suggestion
            FormAction::create('doResetAccount', 'Reset account'),
```

File: src/Extension/AccountReset/SecurityExtension.php (hunk 8 above)
Comment (Cheddam) [reply src/Extension/AccountReset/SecurityExtension.php]:
```suggestion
        $member = Member::get()->byID((int) $memberID);
```

File: src/Extension/AccountReset/SecurityExtension.php (hunk 9 above)
Comment (Cheddam) [reply src/Extension/AccountReset/SecurityExtension.php]: I think the fix for the styling will live in login-forms, yeah. The wording matches the core Reset Password UI, so it's probably best to leave as-is?

**Pull request merged**